lib/spack/spack/compiler.py

```diff
@@ -263,12 +267,17 @@ def _find_matches_in_path(cls, compiler_names, detect_version, *path):
 
                 prod = itertools.product(prefixes, compiler_names, suffixes)
                 for pre, name, suf in prod:
-                    regex = r'^(%s)%s(%s)$' % (pre, re.escape(name), suf)
+                    regex = r'^(%s)(%s)(%s)$' % (pre, re.escape(name), suf)
 
                     match = re.match(regex, exe)
                     if match:
                         key = (full_path,) + match.groups() + (detect_version,)
-                        checks.append(key)
+                        dir_checks.append(key)
+
+            # sort dir_checks by compiler name order
+            # this allows us to prioritize compiler names in subclass
+            dir_checks = sorted(dir_checks, key=check_cmp_key)
```

Comment:
I'd either move the check_cmp_key definition closer to this `sorted()` or just use a `lambda c: compiler_names.index(c[2])` on a new line.  I found it hard to read this currently as what a "check" is doesn't become apparent until you read the loop.

Reply:
@tgamblin refactored.